integration-tests/helpers/fake-agent.js

```diff
@@ -0,0 +1,266 @@
+
+    const messageHandler = msg => {
+      if (msg.payload.request_type !== requestType) return
+      msgCount += 1
+      try {
+        fn(msg)
+        if (msgCount === expectedMessageCount) {
+          resultResolve()
+        }
+      } catch (e) {
+        errors.push(e)
+      }
+      if (msgCount === expectedMessageCount) {
+        this.removeListener('telemetry', messageHandler)
+      }
+    }
+    this.on('telemetry', messageHandler)
+
+    return resultPromise
+  }
+}
+
+function buildExpressServer (agent) {
+  const app = express()
+
+  app.use(bodyParser.raw({ limit: Infinity, type: 'application/msgpack' }))
+  app.use(bodyParser.json({ limit: Infinity, type: 'application/json' }))
+
+  app.put('/v0.4/traces', (req, res) => {
+    if (req.body.length === 0) return res.status(200).send()
+    res.status(200).send({ rate_by_service: { 'service:,env:': 1 } })
+    agent.emit('message', {
+      headers: req.headers,
+      payload: msgpack.decode(req.body, { codec })
+    })
+  })
+
+  app.post('/v0.7/config', (req, res) => {
+    const {
+      client: { products },
+      cached_target_files: cachedTargetFiles
+    } = req.body
+
+    const expires = (new Date(Date.now() + 1000 * 60 * 60 * 24)).toISOString() // in 24 hours
+    const clientID = uuid() // TODO: What is this? It isn't the runtime-id
+
+    // Currently, only `opaque_backend_state` and `targets` are used by dd-trace-js in the object below
+    const targets = {
+      signatures: [],
+      signed: {
+        _type: 'targets',
+        custom: {
+          agent_refresh_interval: 5,
+          opaque_backend_state: ''
+        },
+        expires,
+        spec_version: '1.0.0',
+        targets: {},
+        version: 12345
+      }
+    }
+    const opaqueBackendState = {
+      version: 2,
+      state: { file_hashes: { key: [] } }
+    }
+    const targetFiles = []
+    const clientConfigs = []
+
+    const files = agent._rc_files.filter(({ product }) => products.includes(product))
+
+    for (const { orgId, product, id, name, config } of files) {
+      const path = `datadog/${orgId}/${product}/${id}/${name}`
+      const fileDigest = createHash('sha256').update(config).digest()
+      const fileDigestHex = fileDigest.toString('hex')
+
+      if (cachedTargetFiles.some((cached) =>
+        path === cached.path &&
+        fileDigestHex === cached.hashes.find((e) => e.algorithm === 'sha256').hash
+      )) {
+        continue // skip files already cached by the client so we don't send them more than once
+      }
+
+      opaqueBackendState.state.file_hashes.key.push(fileDigest.toString('base64'))
+
+      targets.signed.targets[path] = {
+        custom: {
+          c: [clientID],
+          'tracer-predicates': { tracer_predicates_v1: [{ clientID }] },
+          v: 20
+        },
+        hashes: { sha256: fileDigestHex },
+        length: config.length
+      }
+
+      targetFiles.push({ path, raw: base64(config) })
+      clientConfigs.push(path)
```

Comment:
pretty sure this needs to be included even when the target_file is cached. So add it before the check. Otherwise if the path is not in client configs, it will consider it needs to delete that config.